File: book-social-network-ui/src/app/service/book-service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { map } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class BookService {

  private BASE_URL = 'http://localhost:8080/books';

  constructor(private http: HttpClient) { }

  // AUTH HEADER
  private authHeaders() {
    const token = localStorage.getItem('token');
    return {
      headers: new HttpHeaders({
        Authorization: `Bearer ${token}`
      })
    };
  }

  // GET ALL BOOKS
  getAll() {
    return this.http.get<any[]>(`${this.BASE_URL}/all`, this.authHeaders());
  }

  // ADD BOOK
  addBook(book: any) {
    return this.http.post(
      `${this.BASE_URL}/create`,
      book,
      this.authHeaders()
    );
  }

  // UPDATE BOOK
  updateBook(id: number, book: any) {
    return this.http.put(
      `${this.BASE_URL}/update/${id}`,
      book,
      this.authHeaders()
    );
  }

  // DELETE BOOK (FIXED)
  borrow(id: number) {
  return this.http.post(`${this.BASE_URL}/borrow/${id}`, null, {
    ...this.authHeaders(),
    responseType: 'text'
  });
}

returnBook(id: number) {
  return this.http.post(`${this.BASE_URL}/return/${id}`, null, {
    ...this.authHeaders(),
    responseType: 'text'
  });
}

deleteBook(id: number) {
  return this.http.delete(`${this.BASE_URL}/delete/${id}`, {
    ...this.authHeaders(),
    responseType: 'text'
  });
}

  // GET BY ID (FIXED)
  getById(id: number) {
    return this.http.get<any>(`${this.BASE_URL}/${id}`, this.authHeaders());
  }

}

File: booksocialnetwork/src/main/java/com/booknetwork/booksocialnetwork/service/BookService.java
package com.booknetwork.booksocialnetwork.service;

import com.booknetwork.booksocialnetwork.dto.BookDto;
import com.booknetwork.booksocialnetwork.entity.Book;

import java.util.List;

public interface BookService {

    BookDto create(BookDto dto, String ownerEmail);

    BookDto getById(Long id);

    BookDto update(Long id, BookDto dto, String ownerEmail);

    void delete(Long id, String ownerEmail);

    BookDto getDtoById(Long id);

    void borrow(Long bookId, String userEmail);

    void returnBook(Long bookId);

    List<BookDto> getAll();

    Book findById(Long id);
}